### PR TITLE
Add OnError

### PR DESCRIPTION
**Pull request by carlmjohnson**
See #52 and #32.

### CODE REVIEW

File: error_handler.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+func ValidationHandler(ok *bool, h ResponseHandler) ErrorHandler {
+	return func(err error, req *http.Request, res *http.Response) {
+		if res != nil && ErrorKindFrom(err) == ErrorKindValidator {
+			*ok = h(res) == nil
+		}
+	}
+}
+
+// JSONError returns an error handler which attempts to decodes a response as a JSON object. If it succeeds, it sets ok to true.
+func JSONError(v any, ok *bool) ErrorHandler {
```
Comment (carlmjohnson) [inline error_handler.go:20]: Maybe ErrorJSON instead.

File: error_handler.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package requests
+
+import (
+	"net/http"
+)
+
+// ErrorHandler is a function accepted by Builder.OnError.
```
Comment (carlmjohnson) [inline error_handler.go:7]: Should note here that req and res might be nil.

File: errorkind.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+func (e ek) Unwrap() error {
+	return e.cause
+}
+
+// ErrorKindFrom extracts the ErrorKind from an error.
+// Nil errors return ErrorKindNone. Errors not passed through
+// a Builder return ErrorKindUnknown.
+func ErrorKindFrom(err error) ErrorKind {
```
Comment (carlmjohnson) [inline errorkind.go:38]: I don't like this name, but the logical name (ErrorKind) is already used by the type. GetErrorKind is ugly. ErrorKindOf is okay, but Of tends to mean "uses generics".

File: error_handler.go (hunk 1 above)
Comment (carlmjohnson) [reply error_handler.go:20]: And add Builder.OnErrorJSON.

File: builder_core.go [hunk 4]
```diff
@@ -178,14 +185,18 @@ func (rb *Builder) Clone() *Builder {
...
+	clip(&rb2.onerrs)
 	return &rb2
 }
 
 // Request builds a new http.Request with its context set.
 func (rb *Builder) Request(ctx context.Context) (req *http.Request, err error) {
 	u, err := url.Parse(rb.baseurl)
 	if err != nil {
-		return nil, fmt.Errorf("could not initialize with base URL %q: %w", u, err)
+		err = fmt.Errorf("could not initialize with base URL %q: %w", u, err)
```
Comment (carlmjohnson) [inline builder_core.go:196]: This message should be shortened.

File: error_handler.go (hunk 1 above)
Comment (carlmjohnson) [reply error_handler.go:20]: Maybe ok should be first to match ValidationHandler?

**Comment by pacoguzman** [comment]
Can I help on moving forward with this PR? I'm very interested on start using it. Thanks

**Comment by carlmjohnson** [comment]
See the discussion in #52. I’m still not convinced the API is right. I can issue an alpha version with an unstable API though if you don’t mind being broken if it changes again.

**Comment by carlmjohnson** [comment]
I [released an alpha version](https://github.com/carlmjohnson/requests/releases/tag/v0.22.4-alpha1). `go get -v -u github.com/carlmjohnson/requests@v0.22.4-alpha1` I would appreciate feedback if you end up using it.

File: errorkind.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+func (e ek) Unwrap() error {
+	return e.cause
+}
+
+// HasKindErr extracts the ErrorKind from an error.
+// Nil errors return KindNone.
+// Errors not from a Builder return KindUnknown.
+func HasKindErr(err error) ErrorKind {
```
Comment (carlmjohnson) [inline errorkind.go:39]: Name doesn't actually work. It should just be `requests.GetErrorKind`.

Comment (carlmjohnson) [reply errorkind.go:39]: I could just drop it now that the handler has changed signatures…

File: builder_core.go [hunk 6]
```diff
@@ -165,6 +166,12 @@ func (rb *Builder) Config(cfgs ...Config) *Builder {
 	return rb
 }
 
+// OnError adds an ErrorHandler to run if any part of building, validating, or handling a request fails.
```
Comment (carlmjohnson) [inline builder_core.go:169]: Add more documentation.

**Comment by pacoguzman** [comment]
Thanks, we will give it a try and provide feedback.

**Comment by carlmjohnson** [comment]
See #63, which has been merged.

**Pull request closed without merging**